### DISCUSSION TITLE
HTML support in hover (inline HTML in IMarkdownString)

### DISCUSSION THREAD

**Discussion by ryanlin1986** [microsoft/monaco-editor#3760] [Q&A]
monaco-editor version: latest
Browser:chrome
OS: win10
Steps or JS usage snippet reproducing the issue:
I want my completion item documentation or hover documentation have inline html.
For example, signature should be colorized, such as:
![image](https://user-images.githubusercontent.com/11850439/38392937-bf21d2f0-395b-11e8-8383-904192e1fc13.png)

Here is what I've tried:
![image](https://user-images.githubusercontent.com/11850439/38392848-7fbeb380-395b-11e8-8c42-608697b320cf.png)
As the above image shows,  the inline html is santinized,.

**Comment by rcjsuen** [comment]
@ryanlin1986 I also opened a separate #802 to describe the problem where the hover behaviour doesn't match VS Code's.

**Comment by mofux** [comment]
Not sure if it's related, but this feature seems to be broken in general since 0.11.x. One good example is if you go to https://microsoft.github.io/monaco-editor/playground.html#interacting-with-the-editor-rendering-glyphs-in-the-margin, switch to the `CSS` tab and hover over one of the rules. The content of the tooltip used to be colorised in previous versions, now it's printed as plain text with strange tick marks around them:

<img width="906" alt="screen shot 2018-04-06 at 01 04 53" src="https://user-images.githubusercontent.com/2785983/38396341-b2427bee-3937-11e8-9434-8d8df9c69071.png">

This might be related to the changes noted in #780, not sure.

**Comment by ryanlin1986** [comment]
@rcjsuen  thanks for the help;)

**Comment by rcjsuen** [comment]
@mofux That's a bug with monaco-css. I've opened #805 as that's a different problem.

**Comment by nikita-fuchs** [comment]
@ryanlin1986 do you happen to have that code example which you used in describing the issue? I have a problem displaying the hover message, as it cannot find `IMarkdownString` for some reason, although it's clearly exported.

**Comment by dehoward** [comment]
@alexdima any update on this bug? 

we're trying to show a `<button>` tag on hover to run a callback. if you have an idea on how to do this using the current implementation I'd love to hear it.

I think we could settle for having a link that _opens in the same tab_ but the current `IMarkdownString` pattern doesn't seem to support this either (just opening links in new tabs).

**Comment by alexdima** [comment]
@dehoward I am not really familiar with the `IMarkdownString`. But I did notice that `IMarkdownString.supportHtml` was added recently via https://github.com/microsoft/vscode/commit/8b7264a91d63da706fb44f9a3dd34cb4c804e7e0 which hints that HTML elements or a subset of HTML elements would be allowed. cc @mjbvz for any extra info.

**Comment by rcjsuen** [comment]
> I think we could settle for having a link that _opens in the same tab_ but the current `IMarkdownString` pattern doesn't seem to support this either (just opening links in new tabs).

I did some experimenting and was able to achieve this via commands inside the `IMarkdownString`. I couldn't get this to work in the [Monaco Editor Playground](https://microsoft.github.io/monaco-editor/playground.html) but maybe it's because it's inside of an `iframe`?

```TypeScript
declare var monaco: any

// create the Monaco editor
const editor = monaco.editor.create(document.getElementById("container")!, {
    language: "plaintext",
    value: "hover here"
});

const cmdId = editor.addCommand(0, () => {
    window.location.href = "https://github.com/microsoft/monaco-editor/";
}, "")

monaco.languages.registerHoverProvider("plaintext", {
    provideHover() {
        return {
            contents: [
                {
                    value: `[Run Command](command:${cmdId})`,
                    isTrusted: true
                },
            ]
        }
    }
});
```

**Comment by rcjsuen** [comment]
@dehoward I figured out a workaround for the `iframe` problem so you can try out the code below in the [Monaco Editor Playground](https://microsoft.github.io/monaco-editor/playground.html).

```JavaScript
monaco.languages.register({ id: 'testLanguage' });

const editor = monaco.editor.create(document.getElementById("container"), {
	value: 'Hover over this text',
	language: 'testLanguage'
});

const cmdId = editor.addCommand(0, () => {
    if (window.self !== window.top) {
        window.parent.location.href = "https://github.com/microsoft/monaco-editor/";;
    } else {
        window.location.href = "https://github.com/microsoft/monaco-editor/";;
    }
})

monaco.languages.registerHoverProvider('testLanguage', {
	provideHover: (model) => {
        return {
            range: new monaco.Range(1, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())),
            contents: [
                { value: `[Run Command](command:${cmdId})`, isTrusted: true }
                
            ]
        };
	}
});
```

**Reply by chen-xiaoxing** [reply]
hi,@rcjsuen,Do you know how to introduce this hover function on demand? Thank you so much

**Reply by rcjsuen** [reply]
Hello, did you mean to use code to make the hover popup? No, I'm not sure how to do that unfortunately.

**Comment by dehoward** [comment]
> @dehoward I am not really familiar with the `IMarkdownString`. But I did notice that `IMarkdownString.supportHtml` was added recently via [microsoft/vscode@8b7264a](https://github.com/microsoft/vscode/commit/8b7264a91d63da706fb44f9a3dd34cb4c804e7e0) which hints that HTML elements or a subset of HTML elements would be allowed. cc @mjbvz for any extra info.

@alexdima Thanks, `IMarkdownString.supportHTML` seems to be what I'm looking for but I haven't been able to get it working in the playground with the below sample:
```js
monaco.languages.register({ id: 'mySpecialLanguage' });

monaco.languages.registerHoverProvider('mySpecialLanguage', {
	provideHover: function (model, position) {
        return {
            range: new monaco.Range(1, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())),
            contents: [
                { value: "<a href='https://microsoft.com'>Link!</a>" },
            ],
            supportHtml: true,
            isTrusted: true
        }
	}
});

monaco.editor.create(document.getElementById("container"), {
	value: '\n\nHover over this text',
	language: 'mySpecialLanguage'
});
```

> @dehoward I figured out a workaround for the iframe problem so you can try out the code below in the Monaco Editor Playground.

@rcjsuen Nice, I think I'll settle for this solution if I'm not able to get `supportHtml` working. Thanks for suggesting!

**Comment by rcjsuen** [comment]
> @alexdima Thanks, `IMarkdownString.supportHTML` seems to be what I'm looking for but I haven't been able to get it working in the playground with the below sample:

@dehoward Your code is a little off because your `supportHtml` and `isTrusted` is in the wrong place. Unfortunately, putting it in the right place seems doesn't seem to help outside of actually rendering it as a hyperlink. May or may not be because of the playground being embedded...?

```JavaScript
monaco.languages.register({ id: 'mySpecialLanguage' });

monaco.languages.registerHoverProvider('mySpecialLanguage', {
	provideHover: function (model) {
        return {
            range: new monaco.Range(1, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())),
            contents: [
                {
                    value: "<a href='https://www.microsoft.com'>Link!</a>",
                    supportHtml: true,
                    isTrusted: true
                },
            ]
        }
	}
});

monaco.editor.create(document.getElementById("container"), {
	value: '\n\nHover over this text',
	language: 'mySpecialLanguage'
});
```

**Comment by alexdima** [comment]
@rcjsuen When I try your snippet, I actually see the correct HTML, so the lack of navigation might be some limitation of the playground's `<iframe>`:
![image](https://user-images.githubusercontent.com/5047891/137035104-228dd6a6-de26-4a1c-aaa3-d78aa981f8e9.png)

**Comment by rcjsuen** [comment]
@alexdima Yeah, I see the same thing. I copy/pasted it to my personal project and it doesn't work there either. I cut my HTML down and it didn't really help.
```HTML
<!DOCTYPE html>
<html>

<head>
    <script src="vs/loader.js"></script>
    <script src="bundle.js"></script>
</head>
<style type="text/css">
    html,
    body {
        overflow: hidden;
        height: 100%;
        margin: 0;
    }

    #container {
        margin: 0;
        border: 0;
        min-width: 100%;
        min-height: 100%;
    }
</style>

<body>
    <div id="container"></div>
</body>

</html>
```

**Comment by dehoward** [comment]
@rcjsuen thanks for catching my mistake. I see the link now with no navigation as you mentioned.  
does a `<button>` work for you? I just see text on my end. 

```js
monaco.languages.register({ id: 'mySpecialLanguage' });

monaco.languages.registerHoverProvider('mySpecialLanguage', {
	provideHover: function (model) {
        return {
            range: new monaco.Range(1, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())),
            contents: [
                {
                    value: "<button onclick='alert(1)'>button!</button>",
                    supportHtml: true,
                    isTrusted: true
                },
            ]
        }
	}
});

monaco.editor.create(document.getElementById("container"), {
	value: '\n\nHover over this text',
	language: 'mySpecialLanguage'
});
```

@alexdima @mjbvz it may be helpful to have some documentation around which "subset of HTML elements" are allowed if there isn't some already.

**Comment by alexdima** [comment]
@dehoward This is not really my area. @mjbvz Can you please help us here?

**Comment by rcjsuen** [comment]
> it may be helpful to have some documentation around which "subset of HTML elements" are allowed if there isn't some already.

@dehoward It's not documented but it's in the [code comments](https://github.com/microsoft/vscode/blob/6d2920473c6f13759c978dd89104c4270a83422d/src/vs/base/browser/markdownRenderer.ts#L292-L296). Neither `input` or `button` is supported.

**Comment by mjbvz** [comment]
Yes that's correct @rcjsuen 

We only allow a safe subset of html tags and attributes (these are not listed directly in the JSDoc for `supportHtml` because that list would likely quickly get out of date). Our safe subset is limited to document-style element, so no buttons or forms. Additionally, you can't use scripts/styles inside the markdown string, so no `onclick` handlers

**Comment by iamjonathanbrown** [comment]
I know this one has been a while, but it looks like the support for rich hover popups is still quite limited? Is there a reason for this limitation? We'd like to switch to using Monaco but it seems like doing the rich popups we do with our existing editor is a nonstarter
